crates/red_knot_python_semantic/resources/mdtest/annotations/literal_string.md

```diff
@@ -0,0 +1,207 @@
+    final: Final[New] = "bar"
+    annotated_class_var: Annotated[
+        ClassVar[Backported], Literal[LiteralString]  # Second arguments and later must be ignored.
+    ] = "foobar"
+
+# TODO: Support new-style generic classes
+# error: [invalid-base]
+class PEP695[L: LiteralString](Protocol):
+    def f[S: Old](self: L | S | New) -> Backported: ...
+    #                   ^^^^^^^^^^^ This is valid, as the class is a protocol.
+
+class GenericParameter(PEP695[LiteralString]): ...
+
+# TODO: Support TypedDict
+class TD(TypedDict):  # error: [invalid-base]
+    normal: LiteralString
+    readonly: ReadOnly[Old]
+    required: Required[New]
+    not_required: NotRequired[Backported]
+```
+
+### Within `Literal`
+
+`LiteralString` cannot be used within `Literal`:
+
+```py
+from typing import Literal, LiteralString
+
+bad_union: Literal["hello", LiteralString]  # error: [invalid-literal-parameter]
+bad_nesting: Literal[LiteralString]  # error: [invalid-literal-parameter]
+```
+
+### Parametrized
+
+`LiteralString` cannot be parametrized.
+
+```py
+# TODO: See above.
+# error: [possibly-unbound-import]
+from typing import LiteralString, TypeAlias, TypeAliasType
+
+Old: TypeAlias = LiteralString
+type New = LiteralString
+Backported = TypeAliasType("Backported", LiteralString)
+
+a: LiteralString[str]  # error: [invalid-type-parameter]
+b: LiteralString["foo"]  # error: [invalid-type-parameter]
+
+c: Old[str]  # error: [invalid-type-parameter]
+d: Old["foo"]  # error: [invalid-type-parameter]
+
+# TODO: Emit invalid-type-parameter for the following
+e: New[str]
+f: New["int"]
+
+g: Backported[str]
+h: Backported["int"]
+
+# fine: TypeAliasType instances are subscriptable.
+# These are not required to have a meaning outside annotation contexts.
+New[str]
+New["int"]
+Backported[str]
+Backported["int"]
+```
+
+### As a base class
+
+Subclassing `LiteralString` leads to a runtime error.
+
+```py
+from typing import LiteralString
+
+class C(LiteralString): ...  # error: [invalid-base]
+```
+
+## Inference
+
+### Common operations
+
+```py
+foo: LiteralString = "foo"
+reveal_type(foo)  # revealed: Literal["foo"]
+
+bar: LiteralString = "bar"
+reveal_type(foo + bar)  # revealed: Literal["foobar"]
+
+baz: LiteralString = "baz"
+baz += foo
+reveal_type(baz)  # revealed: Literal["bazfoo"]
+
+qux = (foo, bar)
+reveal_type(qux)  # revealed: tuple[Literal["foo"], Literal["bar"]]
+
+# TODO: Infer "LiteralString"
+reveal_type(foo.join(qux))  # revealed: @Todo(call type)
+
+template: LiteralString = "{}, {}"
+reveal_type(template)  # revealed: Literal["{}, {}"]
+# TODO: Infer "foo, bar"
+reveal_type(template.format(foo, bar))  # revealed: @Todo(call type)
+```
+
+### Compatibility
+
+`Literal["", ...]` is compatible with `LiteralString`, `LiteralString` is compatible with `str`, but
+not vice versa.
+
+```py
+foo_1: Literal["foo"] = "foo"
+bar_1: LiteralString = foo_1  # fine
+
+if bool():
+    foo_2 = "foo"
+else:
+    foo_2 = "bar"
```

Comment:
This is because you're using `bool()` to create the test, and red-knot knows that `bool()` with no arguments creates an object of type `Literal[False]`. It actually seems like a bug that it _doesn't_ know that for the more verbose form that doesn't use a ternary expression. We shouldn't rely on the bug.

Instead, you should create a custom function that is annotated as returning a `bool`, since red-knot will not be able to "see through" the return annotation, and use _that_ for the test:

```py
def coinflip() -> bool:
    return True

foo_2 = "foo" if coinflip() else "bar"
```